Add FloodForecaster: Domain-Adaptive GINO Framework for Flood Forecasting

MehdiTaghizadehUVa opened the pull request:
<!-- markdownlint-disable MD013-->

# PhysicsNeMo Pull Request

## Description

This PR integrates **FloodForecaster**, a domain-adaptive Geometry-Informed Neural Operator (GINO) framework for rapid, high-resolution flood forecasting. The framework enables accurate, real-time flood predictions by learning from source domain data and adapting to target domains through adversarial training.

### Key Features

- **Domain-Adaptive Training**: Three-stage pipeline (pretraining → domain adaptation → rollout evaluation) for transfer learning from data-rich to data-scarce domains
- **Gradient Reversal Layer (GRL)**: Adversarial domain adaptation using CNN-based domain classifier
- **GINO Integration**: Combines Graph Neural Operators (GNO) and Fourier Neural Operators (FNO) for irregular terrain processing
- **Physics-Informed Metrics**: Volume conservation, arrival time, inundation duration, CSI, and FHCA
- **PhysicsNeMo Module Compliance**: All components inherit from `physicsnemo.Module` with checkpointing support

### Components Added

- Training modules: domain adaptation trainer, pretraining pipeline
- Data processing: GINO wrapper, preprocessing/postprocessing utilities
- Datasets: Custom dataset classes for flood data loading
- Inference: Rollout evaluation and visualization pipeline
- Configuration: Hydra-based configuration system
- Documentation: Complete README with usage examples

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.

## Dependencies

No new dependencies required. All packages are either already in PhysicsNeMo or standard scientific computing libraries:
- `neuralop>=2.0.0` (existing)
- `hydra-core>=1.2.0` (existing)
- `wandb>=0.12.0` (optional, for logging)
- Standard packages: `matplotlib`, `tqdm`, `numpy`, `torch`, `omegaconf`, `pandas`, `h5py`

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

MehdiTaghizadehUVa committed:
Added HydroGraphNet and its corresponding files

MehdiTaghizadehUVa committed:
README file updated.

MehdiTaghizadehUVa committed:
Create animations

MehdiTaghizadehUVa committed:
Add files via upload

MehdiTaghizadehUVa committed:
Delete examples/weather/flood_modeling/hydrographnet/outputs directory

MehdiTaghizadehUVa committed:
Create sample_animation.gif

MehdiTaghizadehUVa committed:
Add files via upload

MehdiTaghizadehUVa committed:
README file updated.

MehdiTaghizadehUVa committed:
Merge remote-tracking branch 'origin/main'

MehdiTaghizadehUVa committed:
README file updated.

MehdiTaghizadehUVa committed:
Update README.md

mnabian committed:
Merge branch 'main' into main

MehdiTaghizadehUVa committed:
Address review comments:
- Moved sample_animation.gif to docs/img
- Removed functions using pickle due to security concerns
- Removed duplicate import of KolmogorovArnoldNetwork in __init__.py

MehdiTaghizadehUVa committed:
Merge remote-tracking branch 'origin/main'

mnabian committed:
formatting

mnabian committed:
Merge branch 'main' into main

MehdiTaghizadehUVa committed:
add unit tests for HydroGraphDataset, KAN layer, and MeshGraphKAN

mnabian committed:
formatting

mnabian committed:
fix meshgraphcan

mnabian committed:
update test data dir

mnabian committed:
fix timeout and tests

mnabian committed:
fix imports

MehdiTaghizadehUVa committed:
fix tests.

MehdiTaghizadehUVa committed:
file for fix tests.

mnabian committed:
Merge branch 'main' into main

mnabian committed:
Merge branch 'main' into main

mnabian committed:
formatting

MehdiTaghizadehUVa committed:
Merge branch 'NVIDIA:main' into main

MehdiTaghizadehUVa committed:
Merge branch 'NVIDIA:main' into main

MehdiTaghizadehUVa committed:
Merge branch 'NVIDIA:main' into main

MehdiTaghizadehUVa committed:
Merge branch 'NVIDIA:main' into main

MehdiTaghizadehUVa committed:
Add FloodForecaster example with domain adaptation

- Add FloodForecaster: domain-adaptive GINO framework for flood forecasting
- Integrate data generation utilities (hydrograph generation, HEC-RAS automation)
- Add comprehensive test coverage for datasets, data processing, training, and utils
- Add rollout animations demonstrating source and target domain performance
- Align code with PhysicsNeMo standards and repository structure

MehdiTaghizadehUVa committed:
Remove data generation utilities and move to separate repository

- Remove data_generation folder and all its contents
- Remove test file for data generation (test_flood_forecaster_data_generation.py)
- Update README.md to reference external repository for data generation
- Remove data generation dependencies from requirements.txt
- Data generation utilities now available at: https://github.com/MehdiTaghizadehUVa/FloodForecaster

MehdiTaghizadehUVa committed:
Rename FloodForecaster folder to floodforecaster - Update test paths for consistency

MehdiTaghizadehUVa committed:
Rename FloodForecaster folder to flood_forecaster for consistency

MehdiTaghizadehUVa committed:
Remove temp_hydrograph.py - accidental error log file

MehdiTaghizadehUVa committed:
Fix all flood_forecaster test errors: query_points, GINOWrapper attributes, da_config mock, gradient reversal, LpLoss signature, and projection mocks

MehdiTaghizadehUVa committed:
Refactor: Separate training and inference, fix all warnings and errors

- Remove rollout evaluation from train.py (moved to inference.py)
- Fix LpLoss warning by wrapping with LpLossWrapper in domain_adaptation
- Fix logger.debug() calls to handle PythonLogger properly
- Fix syntax error in plotting.py (unterminated docstring)
- Fix struct mode error in pretraining.py for neuralop get_model
- Fix error handling in inference.py (exc_info parameter)
- Remove temporary files (GINO_BATCH_SIZE_ANALYSIS.md, 3.0.0)
- Clean separation: train.py for training, inference.py for rollout/visualization

MehdiTaghizadehUVa committed:
Update flood forecaster configuration and training files

MehdiTaghizadehUVa committed:
Fix wandb_step_offset parameter and improve normalizer loading in inference

MehdiTaghizadehUVa committed:
Fix all Greptile review issues: improve code quality, fix bugs, and enhance tests

MehdiTaghizadehUVa committed:
Remove temporary files: PR_DESCRIPTION.md, untrained_checkpoint.mdlus, and CHANGELOG_ENTRY.md

MehdiTaghizadehUVa committed:
Fix test fixture: create train.txt instead of train_.txt to match code fix

mnabian commented on untrained_checkpoint.mdlus:
Please delete this file

mnabian commented on physicsnemo/models/module.py:
```diff
@@ -709,7 +709,7 @@ def load(
 
             # Load state dict after closing archive
-            model_dict = torch.load(io.BytesIO(model_bytes), map_location=device)
+            model_dict = torch.load(io.BytesIO(model_bytes), map_location=device, weights_only=False)
```
Please explain why these changes are needed here. @CharlelieLrt for viz

MehdiTaghizadehUVa replied:
Removed

MehdiTaghizadehUVa committed:
Delete untrained_checkpoint.mdlus

mnabian commented on examples/weather/flood_modeling/flood_forecaster/README.md:
````diff
@@ -0,0 +1,338 @@
+Run the training script:
+
+```bash
+python train.py --config-path conf --config-name config
````
Since you already have these as the default, is it necessary to specify `--config-path` and `--config-name` here?

mnabian commented on examples/weather/flood_modeling/flood_forecaster/README.md:
````diff
@@ -0,0 +1,338 @@
+2. Run the inference script:
+
+```bash
+python inference.py --config-path conf --config-name config
````
Does your inference recipe support multi-GPU/multi-node inference? If yes, please specify the command. If not, please add a note.

mnabian commented on examples/weather/flood_modeling/flood_forecaster/README.md:
```diff
@@ -0,0 +1,338 @@
+- **Jonathan L. Goodall** – <jlg7h@virginia.edu>
+- **Negin Alemazkoor** (Corresponding Author) – <na7fp@virginia.edu>
+
+## License
```
You don't need this section. Please remove it.

mnabian commented on examples/weather/flood_modeling/flood_forecaster/conf/config.yaml:
```diff
@@ -0,0 +1,215 @@
+distributed:
+  use_distributed: false
+  wireup_info: 'mpi'
+  wireup_store: 'tcp'
```
Would you be able to use physicsnemo distributed utilities instead?

mnabian commented on examples/weather/flood_modeling/flood_forecaster/conf/config.yaml:
```diff
@@ -0,0 +1,215 @@
+  save_dir: "./checkpoints_flood_forecaster"
+  resume_from_dir: null
+  resume_from_source: null
+  resume_from_adapt: null
```
Are these needed? If som please describe how these should be used.

mnabian commented on examples/weather/flood_modeling/flood_forecaster/conf/config.yaml:
```diff
@@ -0,0 +1,215 @@
+  training_loss: 'l2'
+  testing_loss: 'l2'
+  weight_decay: 1e-4
+  amp_autocast: false
```
List the available options in a comment

MehdiTaghizadehUVa committed:
Remove weights_only=False from torch.load calls in module.py

MehdiTaghizadehUVa committed:
Clarify that data generation scripts are provided, not the dataset itself

MehdiTaghizadehUVa committed:
Remove redundant Hydra flags from README commands (config-path and config-name are defaults)

MehdiTaghizadehUVa committed:
Address review comments: add config comments, make loss configurable, update distributed docs

MehdiTaghizadehUVa committed:
Migrate FloodForecaster to PhysicsNeMo checkpointing and improve test coverage

- Rename trainer_physicsnemo.py to trainer.py and update all references
- Update GINOWrapper.load_state_dict to handle PhysicsNeMo checkpoint format
- Fix checkpoint loading in inference.py to support both PhysicsNeMo and neuralop formats
- Add comprehensive unit tests for data processing, training, trainer, and inference
- Reduce test suite size to align with other PhysicsNeMo examples
- Remove test coverage analysis markdown file
- Fix various test failures related to model structure and checkpoint loading

MehdiTaghizadehUVa committed:
Add progress bars to training loop and fix autoregressive skip connection

- Add tqdm progress bars for epoch, batch, and evaluation loops
- Fix autoregressive skip connection to handle correct tensor shapes (B, out_channels, n_out) -> (B, n_out, out_channels)
- Add autoregressive config option to model config
- Update pretraining and inference to handle autoregressive parameter correctly

MehdiTaghizadehUVa committed:
Merge branch 'main' into main

MehdiTaghizadehUVa committed:
Fix HydroGraphNet data flow and training correctness

MehdiTaghizadehUVa committed:
Harden FloodForecaster data, training, and checkpoint workflows

MehdiTaghizadehUVa committed:
Restore HydroGraphNet README and correct dataset references

MehdiTaghizadehUVa committed:
Fix FloodForecaster GINO latent-grid wiring

Align the GINOWrapper latent path with native neuralop GINO semantics so mesh-to-grid projection uses latent query points, not mesh coordinates. Also filter loss-only fields from trainer and domain-adaptation model forwards and add regressions covering latent-grid shape handling and target-key filtering.

MehdiTaghizadehUVa committed:
Merge FloodForecaster hardening into main

MehdiTaghizadehUVa committed:
Consolidate FloodForecaster tests and docs

MehdiTaghizadehUVa committed:
Fix FloodForecaster smoke pipeline regressions

MehdiTaghizadehUVa committed:
Restore FloodForecaster README animations

MehdiTaghizadehUVa committed:
Improve FloodForecaster data locality and cache coordination

MehdiTaghizadehUVa committed:
Tune full single-GPU FloodForecaster preset

MehdiTaghizadehUVa committed:
Preserve FloodForecaster resume and scheduler state

MehdiTaghizadehUVa committed:
Apply validated W&B sweep overrides

MehdiTaghizadehUVa committed:
Make FloodForecaster rollout single-rank and asynchronous

MehdiTaghizadehUVa committed:
Validate FloodForecaster feature data contracts

MehdiTaghizadehUVa committed:
Enable fused torch_scatter path for GINO with safe fallback

GINO neighbor reductions ran through neuralop's pure-Python segment_csr
fallback because gno_use_torch_scatter defaulted to false, silently costing
orders of magnitude per training step on GPU. Flip the default to true and
add resolve_gno_scatter_flag() to detect torch_scatter at model-build time:
it keeps the fused path when the extension is present, warns once and
degrades to the slow reduction when it is missing, and warns when the fast
path is available but explicitly disabled. Wire the resolver into both the
training and inference model builders, and declare torch-scatter in
requirements with wheel-index install guidance.

MehdiTaghizadehUVa committed:
Overlap FloodForecaster data loading with GPU compute

The training and domain-adaptation loaders ran with num_workers=0, so every
batch's HDF5 read and tensor construction blocked the GPU on the main thread.
The HDF5 cache backend reopens its file handle per process, so worker forks are
safe. Enable 4 loader workers with persistent workers and pinned host memory to
overlap loading with compute and speed up host-to-device transfers.

MehdiTaghizadehUVa committed:
Guard FloodForecaster eval output logging against empty loaders

The evaluation loop only bound `outs` inside its per-batch body, so an empty
data loader left the name unassigned and the post-loop wandb output-logging
check `outs is not None` raised UnboundLocalError instead of skipping. Initialize
`outs = None` before the loop so the guard degrades cleanly on empty loaders.

MehdiTaghizadehUVa committed:
Add native HDF5 loading for FloodForecaster

MehdiTaghizadehUVa committed:
Document FloodForecaster Zenodo dataset

MehdiTaghizadehUVa committed:
Update FloodForecaster usage documentation
